util/top_gen.py

```diff
@@ -4,11 +4,91 @@
 # SPDX-License-Identifier: Apache-2.0
 
 from mako.template import Template
+import yaml
 
-gpio_num = 5
-uart_num = 5
-i2c_num  = 2
-spi_num  = 7
+# First we parse the YAML input to get the parameters.
+
+with open('data/blocks.yml', 'r') as file:
+    block_parameters = yaml.safe_load(file)
+with open('data/pins.yml', 'r') as file:
+    pin_parameters = yaml.safe_load(file)
+
+# Check block parameter structure.
```

Comment:
We should use an existing YAML schema validation library, e.g. https://pypi.org/project/schema/ (I'm sure there are more there's another one I used in the past but cannot remember what it's called but schema looks pretty decent from a quick glance).

That said for now we should minimise validation work, just putting some asserts in places where you're using the data structure pulled from the YAML to check things are as you expect should do for now. We need to prioritise getting the pinmux generated and usable over good error checking etc.

Reply:
Using Pydantic instead. Thanks to @HU90m for the work.